scripts/get-all-images.py

```diff
@@ -0,0 +1,223 @@
+import argparse
+import logging
+import shutil
+import subprocess
+import sys
+
+import git
+import yaml
+
+# logging
+LOG_FORMAT = "%(levelname)s \t| %(message)s"
+logging.basicConfig(format=LOG_FORMAT, level=logging.INFO)
+log = logging.getLogger(__name__)
+
+# consts
+EXCLUDE_CHARMS = ["kubeflow-roles"]
+
+GH_REPO_KEY = "_github_repo_name"
+GH_BRANCH_KEY = "_github_repo_branch"
+GH_DEPENDENCY_REPO_KEY = "_github_dependency_repo_name"
+GH_DEPENDENCY_BRANCH_KEY = "_github_dependency_repo_branch"
+GET_IMAGES_SH = "tools/get-images.sh"
+
+AIRGAP_TESTING_IMAGES = [
+    "charmedkubeflow/pipelines-runner:ckf-1.8",
+    "docker.io/kubeflowkatib/simple-pbt:v0.16.0",
+    "ghcr.io/knative/helloworld-go:latest",
+    "gcr.io/kubeflow-ci/tf-mnist-with-summaries:1.0",
+]
+
+
+def is_dependency_app(app: dict) -> bool:
+    """
+    Return True if app in bundle is not owned by Analytics team, False
+    otherwise.
+
+    Args:
+        app(dict): app metadata from a bundle.yaml in dictionary form
+
+    Returns:
+        True if app has GH dependency metadata, else False
+    """
+    if GH_DEPENDENCY_REPO_KEY in app and GH_DEPENDENCY_BRANCH_KEY in app:
+        return True
+
+    return False
+
+
+def bundle_app_contains_gh_metadata(app: dict) -> bool:
+    """
+    Given an application in a bundle check if it contains github metadata keys
+    to be able to be parsed properly.
+
+    Args:
+        app(dict): app metadata from a bundle.yaml in dictionary form
+
+    Returns:
+        True if app has GH metadata, else False
+    """
+    if is_dependency_app(app):
+        return True
+
+    if GH_REPO_KEY in app and GH_BRANCH_KEY in app:
+        return True
+
+    return False
+
+
+def clone_repo(repo_name: str, branch: str) -> git.PathLike:
+    """
+    Clones locally a repo and returns the folder created.
+
+    Args:
+        repo_name(str): name of the repo to clone
+        branch(str): branch to checkout to, once cloned the repo
+
+    Returns:
+        The path of the downloaded repository
+    """
+    repo_url = f"https://github.com/canonical/{repo_name}.git"
+
+    logging.info(f"Cloning repo {repo_url}")
+    repo = git.Repo.clone_from(repo_url, str(repo_name))
+
+    logging.info(f"Checking out to branch {branch}")
+    repo.git.checkout(branch)
+
+    return repo.working_dir
+
+
+def get_analytics_app_images(app: dict) -> list[str]:
+    """
+    This function gets the images used by a charm developed by us, by:
+    1. Cloning the repo of the charm
+    2. Running the repo's tools/get-images.sh
+    3. Delete the repo
+
+    If the tools/get-images.sh of a repo fails for any reason then this
+    script will also fail.
+    """
+    images = []
+    repo_dir = clone_repo(app[GH_REPO_KEY], app[GH_BRANCH_KEY])
```

Comment:
Is there reason why not to use [TemporaryDirectory](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory)?

Reply:
Maybe even go further and create custom context manager like this:
```python
@contextmanager
def clone_git_repo(app: dict) -> git.PathLike:
    """Clone repo to tmp directory."""
    repo_name, branch = app[GH_REPO_KEY], app[GH_BRANCH_KEY]

    with tempfile.TemporaryDirectory() as tmp:
        repo_url = f"https://github.com/canonical/{repo_name}.git"

        logging.info(f"Cloning repo {repo_url}")
        repo = git.Repo.clone_from(repo_url, tmp)

        logging.info(f"Checking out to branch {branch}")
        repo.git.checkout(branch)

        yield repo.working_dir
```

Reply:
@rgildein thanks for the suggestion! 

It feels a bit more involved though to use the `TemporaryDirectory` in this case that we only have one file. But I would definitely use this approach in case we had to create a lot of small files and delete them all at once!

Reply:
You do not have one file, you are cloning the whole repo and then removing the folder manually. Using `TemporaryDirectory` will ensure that everything was removed even if there is failure.

Reply:
Unfortunately we won't be able to use the above approach, because of the `yq` snap.

I refactored the code to use the context manager with `TemporaryDirectory`, which means we clone the git repos in paths like `/tmp/tmpfn4ffnhj`. Here's what happens:

1. The repo is cloned under a `tmp/...` tmpdir
2. The python code creates a subprocess and runs the `tools/get-images.sh` script of each repo
3. The above bash script 99% of the time does something like `yq '...' metadata.yaml`
4. Because I'm using the `yq` snap it's in confinement mode, so it can't read things in `/tmp`
5. We see this error `Error: open ./metadata.yaml: no such file or directory`

The above has been reported in the past to `yq` https://github.com/mikefarah/yq/issues/1808

So we'll have to abort this plan, until we convert all the `tools/get-images.sh` scripts to python and remove the dependency of `yq` :/

Reply:
Pushed a commit for this
a8ee551